libraries/Matter/examples/MatterDimmableLight/MatterDimmableLight.ino

```diff
@@ -0,0 +1,171 @@
+// Copyright 2024 Espressif Systems (Shanghai) PTE LTD
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Matter Manager
+#include <Matter.h>
+#include <WiFi.h>
+#include <Preferences.h>
+
+// List of Matter Endpoints for this Node
+// Dimmable Light Endpoint
+#include <MatterEndpoints/MatterDimmableLight.h>
+MatterDimmableLight DimmableLight;
+
+// it will keep last OnOff & Brightness state stored, using Preferences
+Preferences lastStatePref;
+
+// set your board RGB LED pin here
+#ifdef RGB_BUILTIN
+const uint8_t ledPin = RGB_BUILTIN;
+#else
+const uint8_t ledPin = 2;  // Set your pin here if your board has not defined LED_BUILTIN
+#warning "Do not forget to set the RGB LED pin"
+#endif
+
+// set your board USER BUTTON pin here
+const uint8_t buttonPin = 0;  // Set your pin here. Using BOOT Button. C6/C3 use GPIO9.
+
+// WiFi is manually set and started
+const char *ssid = "your-ssid";          // Change this to your WiFi SSID
+const char *password = "your-password";  // Change this to your WiFi password
+
+// Set the RGB LED Light based on the current state of the Dimmable Light
+void setRGBLight(bool state, uint8_t brightness) {
+  Serial.printf("Setting Light to State: %s and Brightness: %d\r\n", DimmableLight ? "ON" : "OFF", brightness);
+  if (state) {
+    rgbLedWrite(ledPin, brightness, brightness, brightness);
+  } else {
+    digitalWrite(ledPin, LOW);
+  }
+}
+
+// Matter Protocol Endpoint On-Off Change Callback
+bool setLightOnOff(bool state) {
+  Serial.printf("User Callback :: New Light State = %s\r\n", state ? "ON" : "OFF");
+  setRGBLight(state, DimmableLight.getBrightness());
+  // store last OnOff state for when the Light is restarted / power goes off
+  lastStatePref.putBool("lastOnOffState", state);
+  // This callback must return the success state to Matter core
+  return true;
+}
+
+// Matter Protocol Endpoint Brightness Change Callback
+bool setLightBrightness(uint8_t brightness) {
+  Serial.printf("User Callback :: New Light Brigthness = %.2f%% [Val = %d]\r\n", ((float) brightness * 100) / MatterDimmableLight::MAX_BRIGHTNESS, brightness);
+  setRGBLight(DimmableLight.getOnOff(), brightness);
+  // store last Brightness for when the Light is restarted / power goes off
+  lastStatePref.putUChar("lastBrightness", brightness);
+  // This callback must return the success state to Matter core
+  return true;
+}
```

Comment:
can't those 3 be one callback?

Reply:
I was thinking about something related yesterday.
There are other 3 light modes to be implemented (temperature, color, all together).

I was considering to use C++ inheritance. Some light modes have the same cluster as the base light and upper lights.

Therefore, keeping the callbacks separated allows me to call parent's methods, reducing the code for the implementation of the children's classes.

Let me implement the other light classes. We can discuss it together looking into the final code.

Reply:
@me-no-dev - I have changed the example code. It makes it simplier. Let me know.

Reply:
@me-no-dev 
It is better to keep the callbacks separated for a simple reason:
Matter protocol allows the controller to change differnt attributes to create effects, like lighting ramp up and down.

Example: A light is ON and level = 50. The user pressed the APP button to turn it OFF. Matter Controller will:
1- Change Level to 1 (keeping it ON) for a ramp down
2- Change on-off to 0 (OFF).
3- Change the Level back to 50, but still keeping it OFF to allow the Light Device to keep/know last brightness (before turning it off).

Therefore, a simple "turn it off" creates  3 different events.
A single callback would be called 3 times but in order to know what has changed (and take action), it would need to compare each attribute with its previous value.

Making it with a callback to each attribute change allows the application to better understand possible individual actions.

Reply:
Last change: used lambda funtions to allow the code to have a single callback function.